Module prepend

archseer opened the pull request:
Things left to port:
- [x] https://github.com/ruby/ruby/commit/8ddbbb33242f413845079b0274c28331cb4aa0f5
- [x] https://github.com/ruby/ruby/commit/f113ab552c7f83bd12f4136bce7fb8ee89819adb
- [x] https://github.com/ruby/ruby/commit/5922c954614e5947a548780bb3b894626affe6dd
- [X] https://github.com/ruby/ruby/commit/59270c607bb4dd053bb4b6ecc0719de011a0eb55
- [X] https://github.com/ruby/ruby/commit/edb1fc4eb2f9b0bd69398362ae40100adfd5577e
- [x] https://github.com/ruby/ruby/commit/a58c224dcff50eec3149df0f91b8a0f2e8099db6
- [ ] https://github.com/ruby/ruby/search?utf8=%E2%9C%93&q=RCLASS_ORIGIN

Commit:
Implement Module#prepend.

Commit:
Rename classes because of conflicts

Commit:
Expose insert position, which should be at origin for include and klass for
prepend.

Commit:
origin must be initialized

Commit:
Comment in a refactor to match MRI, but that fails 320 tests.

IceDragon200 committed:
Additional patches to make this work

archseer commented on src/class.c:
```diff
@@ -759,47 +769,122 @@ boot_defclass(mrb_state *mrb, struct RClass *super)
+MRB_API void
+mrb_include_module(mrb_state *mrb, struct RClass *c, struct RClass *m)
+{
+  include_module_at(mrb, c, c->origin, m, 1);
```
Include must also check `changed` now.

IceDragon200 committed:
Remove some remnant debug code

IceDragon200 committed:
Added changed check to the mrb_include_module

IceDragon200 committed:
Fixed Modules not being fully initialized before #initialize was called

IceDragon200 committed:
Ported a bit more of the MRI Module#prepend tests over

Currently kind_of fails miserably, still looking for the reason

IceDragon200 committed:
Removed some debug prints from the test

IceDragon200 committed:
Ported all MRI prepend tests

And of course, some of them fail miserably

IceDragon200 commented:
@archSeer: Copied over the tests, some of them fail though, give it a run

archSeer commented:
```
Fail: test_prepend_instance_methods_false
 - Assertion[1] Failed: [ruby-dev:45863]
    Expected: [:m1]
      Actual: []
 - Assertion[2] Failed: [ruby-dev:45863]
    Expected: [:m1]
      Actual: []
```

This one requires [this patch](https://bugs.ruby-lang.org/issues/6660):

```diff
diff --git a/class.c b/class.c
index a642ff0..9d346ee 100644
--- a/class.c
+++ b/class.c
@@ -927,7 +927,7 @@ static VALUE
 class_instance_method_list(int argc, VALUE *argv, VALUE mod, int obj, int (*func) (st_data_t, st_data_t, st_data_t))
 {
     VALUE ary;
-    int recur;
+    int recur, prepended = 0;
     st_table *list;
 
     if (argc == 0) {
@@ -939,10 +939,15 @@ class_instance_method_list(int argc, VALUE *argv, VALUE mod, int obj, int (*func
 	recur = RTEST(r);
     }
 
+    if (!recur && RCLASS_ORIGIN(mod) != mod) {
+	mod = RCLASS_ORIGIN(mod);
+	prepended = 1;
+    }
+
     list = st_init_numtable();
     for (; mod; mod = RCLASS_SUPER(mod)) {
 	if (RCLASS_M_TBL(mod)) st_foreach(RCLASS_M_TBL(mod), method_entry_i, (st_data_t)list);
-	if (BUILTIN_TYPE(mod) == T_ICLASS) continue;
+	if (BUILTIN_TYPE(mod) == T_ICLASS && !prepended) continue;
 	if (obj && FL_TEST(mod, FL_SINGLETON)) continue;
 	if (!recur) break;
     }
diff --git a/test/ruby/test_module.rb b/test/ruby/test_module.rb
index d7a6248..1a61535 100644
--- a/test/ruby/test_module.rb
+++ b/test/ruby/test_module.rb
@@ -1299,4 +1299,9 @@ class TestModule < Test::Unit::TestCase
       end
     end
   end
+
+  def test_prepend_instance_methods_fale
+    assert_equal([:m1], Class.new{ prepend Module.new; def m1; end }.instance_methods(false))
+    assert_equal([:m1], Class.new(Class.new{def m2;end}){ prepend Module.new; def m1; end }.instance_methods(false))
+  end
 end

```

Commit:
Include prepended methods in the instance_methods list.

archSeer commented:
Applied patch.

archSeer commented:
The segfault one also seems to run fine for me, `test_prepend_module_ancestors`.

Commit:
Enable visibility prepend tests again

archSeer commented:
There's a whole bunch of code changed inside `vm_method.c`, which we haven't ported: https://github.com/ruby/ruby/commit/8ddbbb33242f413845079b0274c28331cb4aa0f5#diff-6cea7716df08f2b7fdc5fb30c7963884L169

archSeer commented:
@IceDragon200 Also rebased on latest master.

IceDragon200 commented:
@archSeer mruby doesn't seem to have redefine guards, so we won't need the `vm_method.c` changes, the `search_method` is done in `class.c` using `mrb_method_search_vm`, which already has a null check for method tables.

IceDragon200 committed:
Added Module#method_removed hook

IceDragon200 committed:
Removed comment beside method_removed

Not sure if this apart of the ISO standard, so make sure its not misrepresented

IceDragon200 committed:
Bugfix, included_modules would include classes

IceDragon200 committed:
included_modules, origin fix

Prepended modules would include their origin ICLASS

IceDragon200 committed:
Space out test_prepend_super_in_alias assert

Also tried to fix it, however the problem lies with how aliased methods
are done and their internal structure.

mruby simply aliases methods by grabbing the RProc and giving it a new name,
super then determines the original method to call by using the name

so a method called m, aliased as m2, will call the m2 super method instead of m

Commit:
Enable test_prepend_module_ancestors because it seems to pass.

Commit:
Set origin when doing kind_of? comparisons

Commit:
assert() cannot be nested

Commit:
Clean up tests

archseer commented on src/class.c:
```diff
@@ -767,47 +771,132 @@ boot_defclass(mrb_state *mrb, struct RClass *super)
-MRB_API void
-mrb_include_module(mrb_state *mrb, struct RClass *c, struct RClass *m)
+static void
+boot_initmod(mrb_state *mrb, struct RClass *mod)
```
I think this should either be removed, or renamed and extended to function like `class_alloc` in ruby.

IceDragon200 replied:
I'm bad a naming, I tried to make it match the existing `boot_defclass` which is for bootstrapping the first 4 classes (BasicObject, Object, Class and Module). `bootstrap_initialize_module` is what its read as, but yeah, feel free to rename it, its not a class_alloc, since it takes a RClass module and initializes its fields

IceDragon200 replied:
Better yet, we should find all occurences of module creation and run it through one function that does proper initialization, like class

archseer commented on src/class.c:
```diff
@@ -767,47 +771,132 @@ boot_defclass(mrb_state *mrb, struct RClass *super)
-  while (m) {
-    struct RClass *p = c, *ic;
+static struct RClass*
+include_class_new(mrb_state *mrb, struct RClass *m, struct RClass *super)
```
This method is used exactly once.

IceDragon200 replied:
It moves the noisy `ICLASS` creation into a function, this code is very distracting when placed into  `include_module_at`, since its already a complicated function, try describing the function without using too many `and` and you'll see what I mean.

tl;dr separation of concerns

archseer commented on src/class.c:
```diff
@@ -767,47 +771,132 @@ boot_defclass(mrb_state *mrb, struct RClass *super)
+}
+
+MRB_API int
+include_module_at(mrb_state *mrb, struct RClass *klass, struct RClass *c, struct RClass *module, int search_super)
```
We might want to rename these variables to what they were before in mruby. For instance I thought that my way of doing `include_module_at(mrb_state *mrb, struct RClass *c, struct RClass *ins_pos, struct RClass *m, int search_super)` made much more sense, `c` being class, `ins_pos`/`cur` being the position where modules are inserted and `m` being the module that is being included.

IceDragon200 committed:
Renamed parameters in include_module_at

IceDragon200 committed:
Applied gc patch to fix ORIGIN ICLASS method table leak

Based on the gc patch by ko1

https://github.com/ruby/ruby/commit/5922c954614e5947a548780bb3b894626affe6dd

IceDragon200 committed:
Remove non-applicable "hack" comment

IceDragon200 committed:
Renamed MRB_FLAG_IS_INSTANCE to MRB_INSTANCE_TT_MASK

IceDragon200 committed:
Make include_module_at static

Since I can't forsee any reason to use it directly inplace of using
prepend/include

IceDragon200 committed:
Added mrb_prepend_module to mruby header
